# Food Standards Agency 

The United Kingdom's Food Standards Agency (FSA) is responsible for maintaining standards in food safety in England, Wales and Northern Ireland, in part through the enforcement of food related regulations. (Scotland operates a slightly different scheme.)



- The Food Hygiene Rating Scheme (FHRS) data is licensed under the Open Government Licence v3.0, which allows users to copy, publish, distribute, and transmit the data, to adapt the data ,and to use the data commercially.
- The food hygiene rating data can be collected via an application programming interface (API) in XML and JSON formats. Also the Data Catalog section contain datasets which can be downloaded in csv and pdf formats.
- Establishment Name and ID, Establishment address, Establishment postcode, Business type (category), Inspection date, Scope, Intervention rating score for level of compliance - food hygiene and safety procedures, Intervention rating score for level of compliance – structure of the establishment, Intervention rating score for confidence in management/control procedures, Revised food hygiene
rating, and Revised food hygiene rating date.
- A positive aspect is that the data is free to use and promotes transparency. However, a drawback is that the data may not accurately reflect the current status of establishments due to infrequent updates.


In [2]:
# We'll load a couple of libraries that we'll be using in this Notebook.
# 'requests' is an HTTP library - it allows us to get data from a URL address
import pandas as pd
import json
import requests

In [3]:
url = "https://ratings.food.gov.uk/api/open-data-files/FHRS870en-GB.json"

# Make a query to the URL to retrieve some JSON data

fsa_resp = requests.get(url)

fsa = fsa_resp.json()

json.dumps(fsa)[:1000]

'{"FHRSEstablishment": {"Header": {"ExtractDate": "2025-02-23", "ItemCount": 2324, "ReturnCode": "Success"}, "EstablishmentCollection": [{"FHRSID": 1293318, "LocalAuthorityBusinessID": "91816", "BusinessName": "12th Street Burgers, Shakes & Dynamite Wings", "BusinessType": "Restaurant/Cafe/Canteen", "BusinessTypeID": 1, "AddressLine2": "16 Garrick Walk", "AddressLine3": "Central Milton Keynes", "AddressLine4": "Milton Keynes", "PostCode": "MK9 3PT", "RatingValue": "5", "RatingKey": "fhrs_5_en-GB", "RatingDate": "2023-11-17", "LocalAuthorityCode": "870", "LocalAuthorityName": "Milton Keynes", "LocalAuthorityWebSite": "http://www.milton-keynes.gov.uk", "LocalAuthorityEmailAddress": "ehfst@milton-keynes.gov.uk", "Scores": {"Hygiene": 0, "Structural": 5, "ConfidenceInManagement": 0}, "SchemeType": "FHRS", "NewRatingPending": false, "Geocode": {"Longitude": "-0.7501625", "Latitude": "52.0435764"}}, {"FHRSID": 1285693, "LocalAuthorityBusinessID": "91784", "BusinessName": "14Forty", "Business

In [4]:
type(fsa)

dict

We can see here that `EstablishmentCollection` contains the headers which we need. 

Now we will use pandas library to load `EstablishmentCollection` keys by using json_normalize() function which provides a range of arguments that allow you to customise how a particular JSON parsed data structure is treated.

In [5]:
fsa = fsa["FHRSEstablishment"]['EstablishmentCollection']
df_fsa = pd.json_normalize(fsa)
df_fsa.head()

FHRSID LocalAuthorityBusinessID  \
0  1293318                    91816   
1  1285693                    91784   
2   198797                    64942   
3   135218                    68752   
4  1776134                    96015   

                                   BusinessName             BusinessType  \
0  12th Street Burgers, Shakes & Dynamite Wings  Restaurant/Cafe/Canteen   
1                                       14Forty  Restaurant/Cafe/Canteen   
2                                      1st Stop        Retailers - other   
3                                      360 Play  Restaurant/Cafe/Canteen   
4                                4ster's Treats  Other catering premises   

   BusinessTypeID      AddressLine2           AddressLine3   AddressLine4  \
0               1   16 Garrick Walk  Central Milton Keynes  Milton Keynes   
1               1  Pendeen Crescent          Milton Keynes  Milton Keynes   
2            4613     125 Queensway              Bletchley  Milton Keynes   
3               1       Roebuck Way              Knowlhill  Milton Keynes   
4            7841               NaN                    NaN            NaN   

  PostCode         RatingValue  ... SchemeType NewRatingPending  \
0  MK9 3PT                   5  ...       FHRS            False   
1  MK4 4BX                   4  ...       FHRS            False   
2  MK2 2DH                   5  ...       FHRS            False   
3  MK5 8HL                   5  ...       FHRS            False   
4      NaN  AwaitingInspection  ...       FHRS            False   

  Scores.Hygiene Scores.Structural Scores.ConfidenceInManagement  \
0            0.0               5.0                           0.0   
1            0.0              10.0                           5.0   
2            5.0               0.0                           5.0   
3            5.0               0.0                           5.0   
4            NaN               NaN                           NaN   

  Geocode.Longitude Geocode.Latitude  AddressLine1  Geocode  Scores  
0        -0.7501625       52.0435764           NaN      NaN     NaN  
1               NaN              NaN  DHL Nautilus      NaN     NaN  
2          -0.72897        51.995092           NaN      NaN     NaN  
3         -0.771821        52.021711      360 Play      NaN     NaN  
4               NaN              NaN           NaN      NaN     NaN  

[5 rows x 26 columns]

In [7]:
df_fsa.tail()

FHRSID LocalAuthorityBusinessID         BusinessName  \
2319  1563660                    94199                Zayka   
2320  1739018                    91369  Zeenath Food Centre   
2321   137094                    71779                Zizzi   
2322  1700962                    95271         Zorro's Cafe   
2323   615498                    83891          Zouk Lounge   

                 BusinessType  BusinessTypeID       AddressLine2  \
2319  Other catering premises            7841   226 Burney Drive   
2320        Retailers - other            4613             Dodkin   
2321  Restaurant/Cafe/Canteen               1            The Hub   
2322           Mobile caterer            7846                NaN   
2323  Restaurant/Cafe/Canteen               1  34 Stratford Road   

               AddressLine3   AddressLine4  PostCode RatingValue  ...  \
2319             Glebe Farm  Milton Keynes  MK17 8LE           5  ...   
2320               Beanhill  Milton Keynes   MK6 4LP           3  ...   
2321  Central Milton Keynes  Milton Keynes   MK9 2HP           5  ...   
2322                    NaN            NaN       NaN           4  ...   
2323              Wolverton  Milton Keynes  MK12 5LW           3  ...   

     SchemeType NewRatingPending Scores.Hygiene Scores.Structural  \
2319       FHRS            False            0.0               0.0   
2320       FHRS            False           10.0              10.0   
2321       FHRS            False            0.0               0.0   
2322       FHRS            False            5.0              10.0   
2323       FHRS            False            5.0              10.0   

     Scores.ConfidenceInManagement   Geocode.Longitude  Geocode.Latitude  \
2319                           5.0          -0.6670064        52.0322111   
2320                          10.0          -0.7358228        52.0179505   
2321                           0.0  -0.764765977859497  52.0392684936523   
2322                           5.0                 NaN               NaN   
2323                          10.0            -0.81205         52.062763   

                                           AddressLine1  Geocode  Scores  
2319                                                NaN      NaN     NaN  
2320                                             Unit 3      NaN     NaN  
2321                                                NaN      NaN     NaN  
2322  Trading Location Central Milton Keynes Market ...      NaN     NaN  
2323                                                NaN      NaN     NaN  

[5 rows x 26 columns]

checking the keys and the columns

In [5]:
df_fsa.columns, fsa[0].keys()

(Index(['FHRSID', 'LocalAuthorityBusinessID', 'BusinessName', 'BusinessType',
        'BusinessTypeID', 'AddressLine2', 'AddressLine3', 'AddressLine4',
        'PostCode', 'RatingValue', 'RatingKey', 'RatingDate',
        'LocalAuthorityCode', 'LocalAuthorityName', 'LocalAuthorityWebSite',
        'LocalAuthorityEmailAddress', 'SchemeType', 'NewRatingPending',
        'Scores.Hygiene', 'Scores.Structural', 'Scores.ConfidenceInManagement',
        'Geocode.Longitude', 'Geocode.Latitude', 'AddressLine1', 'Geocode',
        'Scores'],
       dtype='object'),
 dict_keys(['FHRSID', 'LocalAuthorityBusinessID', 'BusinessName', 'BusinessType', 'BusinessTypeID', 'AddressLine2', 'AddressLine3', 'AddressLine4', 'PostCode', 'RatingValue', 'RatingKey', 'RatingDate', 'LocalAuthorityCode', 'LocalAuthorityName', 'LocalAuthorityWebSite', 'LocalAuthorityEmailAddress', 'Scores', 'SchemeType', 'NewRatingPending', 'Geocode']))

It is quite clear that we can not see all of the columns contained in the dataframe. The dataset  contain nested information which is not appropriate to be represented in such format.  

Also We can notice that the `Geocode` and `Scores` columns are empty because they are designed to function as keys that reference nested information, causing such columns to be empty when using undocumented methods.

| This attribute                                               | functionally defines this attribute |
| ------------------------------------------------------------ | :---------------------------------- |
| `FHRSID`                                                 | `LocalAuthorityBusinessID`, `BusinessName`, `BusinessTypeID`, `AddressLine1`, `AddressLine2`, `AddressLine3`, `AddressLine4`, `PostCode`, `RatingKey`, `RatingDate`, `LocalAuthorityCode`, `NewRatingPending`, `Scores.Hygiene`, `Scores.Structural`, `Scores.ConfidenceInManagement`, `Geocode.Longitude`, `Geocode.Latitude`                     |
| `LocalAuthorityCode`                                                 | `LocalAuthorityName, LocalAuthorityWebSite, LocalAuthorityEmailAddress`                         |
| `BusinessTypeID`                                                  | `BusinessType`                       |
| `RatingKey`                                                  | `RatingValue`                       |


Based on the cell above i would have 4 tables.  

The first table would be representing **Businesses**.  
The columns in this table would be `FHRSID`, `LocalAuthorityBusinessID`, `BusinessName`, `BusinessTypeID`, `AddressLine1`, `AddressLine2`, `AddressLine3`, `AddressLine4`, `PostCode`, `RatingKey`, `RatingDate`, `LocalAuthorityCode`, `NewRatingPending`, `Scores.Hygiene`, `Scores.Structural`, `Scores.ConfidenceInManagement`, `Geocode.Longitude`, `Geocode.Latitude`  

`FHRSID` would be the Primary Key, which must be unique and not NULL.


The Foreign keys are:  
`LocalAuthorityCode` which references `LocalAuthorities.LocalAuthorityCode`  
`BusinessTypeID` which references `BusinessTypes.BusinessTypeID`  
`RatingKey` which references `Ratings.RatingKey`




The second table would be representing the **Local Authorities**.  

The columns in this table would be `LocalAuthorityCode`, `LocalAuthorityName, LocalAuthorityWebSite, LocalAuthorityEmailAddress`   
The primary key is `LocalAuthorityCode`, which must be unique and not NULL.  





The third table would be representing the __Business Type__.  

The columns are  `BusinessTypeID` as the primary Key, which must be unique and not NULL, and `BusinessType`

The final table would be representing the **Ratings**.  

The columns are `RatingKey` as the primary Key, which must be unique and not NULL, and `RatingValue`

In [2]:
MONGO_CONNECTION_STRING = f"mongodb://localhost:27017/"
print(f"MONGO_CONNECTION_STRING = {MONGO_CONNECTION_STRING}")

MONGO_CONNECTION_STRING = mongodb://localhost:27017/


In [3]:
from pymongo import MongoClient

In [5]:
mongo_client = MongoClient(MONGO_CONNECTION_STRING)

In [6]:
DB_NAME = "fsa"  
print(f"DB_NAME = {DB_NAME}")

mongo_db = mongo_client[DB_NAME]

DB_NAME = fsa


In [11]:
! wget https://ratings.food.gov.uk/api/open-data-files/FHRS870en-GB.json  -O fsa_data.json

--2025-02-13 15:44:52--  https://ratings.food.gov.uk/api/open-data-files/FHRS870en-GB.json
Resolving wwwcache.open.ac.uk (wwwcache.open.ac.uk)... 172.28.58.12
Connecting to wwwcache.open.ac.uk (wwwcache.open.ac.uk)|172.28.58.12|:80... connected.
Proxy request sent, awaiting response... 200 OK
Length: 1526523 (1.5M) [application/json]
Saving to: ‘fsa_data.json’

fsa_data.json       100%[===================>]   1.46M  --.-KB/s    in 0.05s   

2025-02-13 15:44:55 (31.4 MB/s) - ‘fsa_data.json’ saved [1526523/1526523]



In [12]:
import json

with open("fsa_data.json") as f:
    data = json.load(f)

data["FHRSEstablishment"]["Header"]["ItemCount"], data["FHRSEstablishment"]["EstablishmentCollection"][:1]

(2324,
 [{'FHRSID': 1293318,
   'LocalAuthorityBusinessID': '91816',
   'BusinessName': '12th Street Burgers, Shakes & Dynamite Wings',
   'BusinessType': 'Restaurant/Cafe/Canteen',
   'BusinessTypeID': 1,
   'AddressLine2': '16 Garrick Walk',
   'AddressLine3': 'Central Milton Keynes',
   'AddressLine4': 'Milton Keynes',
   'PostCode': 'MK9 3PT',
   'RatingValue': '5',
   'RatingKey': 'fhrs_5_en-GB',
   'RatingDate': '2023-11-17',
   'LocalAuthorityCode': '870',
   'LocalAuthorityName': 'Milton Keynes',
   'LocalAuthorityWebSite': 'http://www.milton-keynes.gov.uk',
   'LocalAuthorityEmailAddress': 'ehfst@milton-keynes.gov.uk',
   'Scores': {'Hygiene': 0, 'Structural': 5, 'ConfidenceInManagement': 0},
   'SchemeType': 'FHRS',
   'NewRatingPending': False,
   'Geocode': {'Longitude': '-0.7501625', 'Latitude': '52.0435764'}}])

We can observe the total number of records is 2324. These records are stored in a list object accessed by the `EstablishmentCollection` key, which is stored in the `FHRSEstablishment` key.  



First, we will delete any existing instances of the food_ratings_raw  collection 

In [13]:
mongo_db.drop_collection("food_ratings_raw")

{'nIndexesWas': 1, 'ns': 'fsa.food_ratings_raw', 'ok': 1.0}

Then we create the collection

In [14]:
food_ratings_raw = mongo_db["food_ratings_raw"]

In [7]:
# Check the database object
mongo_db

Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'fsa')

Now we add the records to the collection and then check one record.

In [17]:
food_ratings_raw.insert_many(data["FHRSEstablishment"]["EstablishmentCollection"])

food_ratings_raw.find_one()

{'_id': ObjectId('67ae138932182a780d58c6a0'),
 'FHRSID': 1293318,
 'LocalAuthorityBusinessID': '91816',
 'BusinessName': '12th Street Burgers, Shakes & Dynamite Wings',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'AddressLine2': '16 Garrick Walk',
 'AddressLine3': 'Central Milton Keynes',
 'AddressLine4': 'Milton Keynes',
 'PostCode': 'MK9 3PT',
 'RatingValue': '5',
 'RatingKey': 'fhrs_5_en-GB',
 'RatingDate': '2023-11-17',
 'LocalAuthorityCode': '870',
 'LocalAuthorityName': 'Milton Keynes',
 'LocalAuthorityWebSite': 'http://www.milton-keynes.gov.uk',
 'LocalAuthorityEmailAddress': 'ehfst@milton-keynes.gov.uk',
 'Scores': {'Hygiene': 0, 'Structural': 5, 'ConfidenceInManagement': 0},
 'SchemeType': 'FHRS',
 'NewRatingPending': False,
 'Geocode': {'Longitude': '-0.7501625', 'Latitude': '52.0435764'}}

Lets see how many documents we now have in the collection

In [18]:
food_ratings_raw.count_documents({})

2324

As expected we have 2324 records which matches the number of establishments listed in the original data.

In [ ]:
partial_validation_schema = {
    "$jsonSchema": {
        "bsonType": "object",
        "properties": {
            "Geocode": {
                "bsonType": ["object", "null"],
                "properties": {
                    "Longitude": {
                        "bsonType": "double",
                        "description": "if present, must be a double",
                    },
                    "Latitude": {
                        "bsonType": "double",
                        "description": "if present, must be a double",
                    },
                },
            }
        },
    }
}

First, we will delete any existing instances of the `food_ratings_cleaner` collection

In [20]:
mongo_db.drop_collection("food_ratings_cleaner")

{'nIndexesWas': 1, 'ns': 'fsa.food_ratings_cleaner', 'ok': 1.0}

In [21]:
VALIDATED_COLLECTION =  "food_ratings_cleaner"

food_ratings_cleaner = mongo_db.create_collection(VALIDATED_COLLECTION, validator=partial_validation_schema, validationLevel="strict" )
food_ratings_cleaner

Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'fsa'), 'food_ratings_cleaner')

In [22]:
food_ratings_cleaner.insert_many(data["FHRSEstablishment"]["EstablishmentCollection"])

BulkWriteError: batch op errors occurred, full error: {'writeErrors': [{'index': 0, 'code': 121, 'errmsg': 'Document failed validation', 'errInfo': {'failingDocumentId': ObjectId('67ae138932182a780d58c6a0'), 'details': {'operatorName': '$jsonSchema', 'schemaRulesNotSatisfied': [{'operatorName': 'properties', 'propertiesNotSatisfied': [{'propertyName': 'Geocode', 'details': [{'operatorName': 'properties', 'propertiesNotSatisfied': [{'propertyName': 'Longitude', 'description': 'if present, must be a double', 'details': [{'operatorName': 'bsonType', 'specifiedAs': {'bsonType': 'double'}, 'reason': 'type did not match', 'consideredValue': '-0.7501625', 'consideredType': 'string'}]}, {'propertyName': 'Latitude', 'description': 'if present, must be a double', 'details': [{'operatorName': 'bsonType', 'specifiedAs': {'bsonType': 'double'}, 'reason': 'type did not match', 'consideredValue': '52.0435764', 'consideredType': 'string'}]}]}]}]}]}}, 'op': {'FHRSID': 1293318, 'LocalAuthorityBusinessID': '91816', 'BusinessName': '12th Street Burgers, Shakes & Dynamite Wings', 'BusinessType': 'Restaurant/Cafe/Canteen', 'BusinessTypeID': 1, 'AddressLine2': '16 Garrick Walk', 'AddressLine3': 'Central Milton Keynes', 'AddressLine4': 'Milton Keynes', 'PostCode': 'MK9 3PT', 'RatingValue': '5', 'RatingKey': 'fhrs_5_en-GB', 'RatingDate': '2023-11-17', 'LocalAuthorityCode': '870', 'LocalAuthorityName': 'Milton Keynes', 'LocalAuthorityWebSite': 'http://www.milton-keynes.gov.uk', 'LocalAuthorityEmailAddress': 'ehfst@milton-keynes.gov.uk', 'Scores': {'Hygiene': 0, 'Structural': 5, 'ConfidenceInManagement': 0}, 'SchemeType': 'FHRS', 'NewRatingPending': False, 'Geocode': {'Longitude': '-0.7501625', 'Latitude': '52.0435764'}, '_id': ObjectId('67ae138932182a780d58c6a0')}}], 'writeConcernErrors': [], 'nInserted': 0, 'nUpserted': 0, 'nMatched': 0, 'nModified': 0, 'nRemoved': 0, 'upserted': []}

The BulkWriteError occurs because the Longitude and Latitude values in the Geocode field are stored as strings instead of doubles as required.

In [24]:
# Uncomment the following cell to create a backup of your database
# in the directory: mongo_backups
! mongodump --db=fsa --out=./mongo_backups

2025-02-13T15:47:28.560+0000	writing fsa.food_ratings_cleaner_validated_connection to mongo_backups/fsa/food_ratings_cleaner_validated_connection.bson
2025-02-13T15:47:28.560+0000	writing fsa.food_ratings_raw to mongo_backups/fsa/food_ratings_raw.bson
2025-02-13T15:47:28.563+0000	done dumping fsa.food_ratings_cleaner_validated_connection (0 documents)
2025-02-13T15:47:28.563+0000	writing fsa.food_ratings_cleaner to mongo_backups/fsa/food_ratings_cleaner.bson
2025-02-13T15:47:28.570+0000	done dumping fsa.food_ratings_raw (2324 documents)
2025-02-13T15:47:28.573+0000	done dumping fsa.food_ratings_cleaner (2324 documents)


In [29]:
# Drop the fsa database and collections
mongo_client.drop_database(DB_NAME)
mongo_client.list_database_names(), mongo_db.list_collection_names()

(['admin', 'config', 'local'], [])

Reloading the database

In [30]:
! mongorestore ./mongo_backups

2025-02-13T15:51:46.702+0000	preparing collections to restore from
2025-02-13T15:51:46.703+0000	reading metadata for fsa.food_ratings_cleaner from mongo_backups/fsa/food_ratings_cleaner.metadata.json
2025-02-13T15:51:46.703+0000	reading metadata for fsa.food_ratings_cleaner_validated_connection from mongo_backups/fsa/food_ratings_cleaner_validated_connection.metadata.json
2025-02-13T15:51:46.703+0000	reading metadata for fsa.food_ratings_raw from mongo_backups/fsa/food_ratings_raw.metadata.json
2025-02-13T15:51:46.782+0000	restoring fsa.food_ratings_raw from mongo_backups/fsa/food_ratings_raw.bson
2025-02-13T15:51:46.797+0000	restoring fsa.food_ratings_cleaner from mongo_backups/fsa/food_ratings_cleaner.bson
2025-02-13T15:51:46.813+0000	restoring fsa.food_ratings_cleaner_validated_connection from mongo_backups/fsa/food_ratings_cleaner_validated_connection.bson
2025-02-13T15:51:46.823+0000	finished restoring fsa.food_ratings_cleaner_validated_connection (0 documents, 0 failures)
2025-02

In [63]:
# Testing the restored database
mongo_client.list_database_names(), mongo_db.list_collection_names(), food_ratings_cleaner.find_one()

(['admin', 'config', 'fsa', 'local'],
 ['food_ratings_raw',
  'food_ratings_cleaner',
  'food_ratings_cleaner_validated_connection'],
 {'_id': ObjectId('67ae138932182a780d58c6a0'),
  'FHRSID': 1293318,
  'LocalAuthorityBusinessID': '91816',
  'BusinessName': '12th Street Burgers, Shakes & Dynamite Wings',
  'BusinessType': 'Restaurant/Cafe/Canteen',
  'BusinessTypeID': 1,
  'AddressLine2': '16 Garrick Walk',
  'AddressLine3': 'Central Milton Keynes',
  'AddressLine4': 'Milton Keynes',
  'PostCode': 'MK9 3PT',
  'RatingValue': '5',
  'RatingKey': 'fhrs_5_en-GB',
  'RatingDate': '2023-11-17',
  'LocalAuthorityCode': '870',
  'LocalAuthorityName': 'Milton Keynes',
  'LocalAuthorityWebSite': 'http://www.milton-keynes.gov.uk',
  'LocalAuthorityEmailAddress': 'ehfst@milton-keynes.gov.uk',
  'Scores': {'Hygiene': 0, 'Structural': 5, 'ConfidenceInManagement': 0},
  'SchemeType': 'FHRS',
  'NewRatingPending': False,
  'Geocode': {'Longitude': -0.7501625, 'Latitude': 52.0435764}})

In [109]:
import folium

First i need to check the different types of Businesses

In [110]:
df_fsa['BusinessType'].unique()

array(['Restaurant/Cafe/Canteen', 'Retailers - other',
       'Other catering premises', 'Hospitals/Childcare/Caring Premises',
       'Mobile caterer', 'School/college/university',
       'Takeaway/sandwich shop', 'Retailers - supermarkets/hypermarkets',
       'Pub/bar/nightclub', 'Manufacturers/packers',
       'Hotel/bed & breakfast/guest house', 'Distributors/Transporters',
       'Farmers/growers', 'Importers/Exporters'], dtype=object)

Next we check what the result will look like

In [117]:
food_ratings_cleaner.find_one({"BusinessName": {"$regex" : ".*Premier.*", "$options": "i"}, "LocalAuthorityName" : "Milton Keynes"}, {"BusinessName" :1, "BusinessType" : 1,
"Geocode" : 1})

{'_id': ObjectId('67ae138932182a780d58cca7'),
 'BusinessName': 'Premier',
 'BusinessType': 'Retailers - other',
 'Geocode': {'Longitude': -0.6834376, 'Latitude': 52.0431983}}

In [119]:
pd.DataFrame(food_ratings_cleaner.find({"BusinessName": {"$regex" : ".*Premier.*", "$options": "i"}, "BusinessType" : "Retailers - other",  "LocalAuthorityName" : "Milton Keynes"}, {"BusinessName" :1,"Geocode" : 1 }, limit=5))

_id                         BusinessName  \
0  67ae138932182a780d58cca7                              Premier   
1  67ae138932182a780d58cca8                              Premier   
2  67ae138932182a780d58cca9                              Premier   
3  67ae138932182a780d58ccaa  Premier - Church Street Mini Market   
4  67ae138932182a780d58ccb9              Premier Newport Pagnell   

                                             Geocode  
0  {'Longitude': -0.6834376, 'Latitude': 52.0431983}  
1    {'Longitude': -0.808236, 'Latitude': 52.052618}  
2    {'Longitude': -0.741492, 'Latitude': 52.032117}  
3    {'Longitude': -0.811671, 'Latitude': 52.062211}  
4    {'Longitude': -0.722494, 'Latitude': 52.086957}

So we need to load this in a Data Frame and we need to separate  the longitude and the latitude. We can do this by using the json_normalize method.

In [138]:
premier_locations = food_ratings_cleaner.find({"BusinessName": {"$regex" : ".*Premier.*"}, "BusinessType" : "Retailers - other",  "LocalAuthorityName" : "Milton Keynes"}, {"BusinessName" :1,"Geocode" : 1, "_id" :0})
premier_locations = pd.json_normalize(list(premier_locations))

In [139]:
premier_locations.head()

BusinessName  Geocode.Longitude  Geocode.Latitude
0                              Premier          -0.683438         52.043198
1                              Premier          -0.808236         52.052618
2                              Premier          -0.741492         52.032117
3  Premier - Church Street Mini Market          -0.811671         52.062211
4              Premier Newport Pagnell          -0.722494         52.086957

This looks okay but it's better to rename the two columns to simply longitude and latitude. 

In [140]:
premier_locations.rename(columns={"Geocode.Longitude" : "Longitude", "Geocode.Latitude" :"Latitude"}, inplace=True)
premier_locations.head()

BusinessName  Longitude   Latitude
0                              Premier  -0.683438  52.043198
1                              Premier  -0.808236  52.052618
2                              Premier  -0.741492  52.032117
3  Premier - Church Street Mini Market  -0.811671  52.062211
4              Premier Newport Pagnell  -0.722494  52.086957

Next we need to make the map so we will take the median of the locations and create our map. 

In [141]:
AVERAGE_LOCATION = premier_locations[["Latitude", "Longitude"]].median()

In [151]:
def add_marker(row, m):
    folium.Circle(
        location=[row["Latitude"], row["Longitude"]],
        color="red",
        radius=50,
        fill=True,
        fill_opacity=1.0,
        popup=folium.Popup(f"{row['BusinessName']}")
 
    ).add_to(m)

In [152]:
m = folium.Map(tuple(AVERAGE_LOCATION), width=500, height=700, zoom_start=10)

premier_locations.apply(add_marker, m=m, axis=1)

m

In [7]:
aldi_fn = "2024J_TMA02_data/FSA/fsa_establishments_aldi_03-07-2024.json"

# Add Data From JSON
import json

with open(aldi_fn) as f:
    mongo_db["food_ratings_aldi"].insert_many(
        json.load(f)["EstablishmentCollection"]["EstablishmentDetail"]
    )

In [9]:
# Check the key we need to match on in the geojson file
from IPython.display import JSON

geojson_fn = "2024J_TMA02_data/Boundaries/LAD_boundaries_BUC_4326_2023.geojson"

# Uncomment the following line before running this cell
# to preview the geoJSON data structure
# Navigate into one of the `features` and inspect the `properties`
JSON(geojson_fn)

<IPython.core.display.JSON object>

For this specific dataset, a document database is a better fit than a relational database because it handles nested documents like the `Geocode` keys with values such as `latitude` and `longitude`, storing them as a single entity rather making it much easier to retrieve and to work with the data. It also handles missing values such as the absent `latitude` and `longitude` of some establishments, allowing the key `Geocode` not appear at all since it becomes irrelevant.  

Now when i think of a disadvantage i think of ACID which made me notice keys like `LocalAuthorityName` which duplicated across multiple documents in the data risking inconsistencies if an update or a change is required.  
In a relational database, we can normalize this data into a separate `LocalAuthorities` table, linked to businesses via foreign keys. This reduces redundancy and ensures consistency.   


However while relational databases offer follow ACID, the flexibility and efficiency of a document database such as MongoDB are more valuable for this dataset, where nested structures and Null fields are common.